### PR TITLE
TLSv1.3: Disallow compression

### PR DESCRIPTION
**Pull request by mattcaswell**
##### Checklist

- [x] tests are added or updated

##### Description of change

Make sure that we can never negotiate compression in TLSv1.3. As per the spec this means that even if the server only support TLSv1.2, the client will never negotiate compression if it issues a TLSv1.3 compliant ClientHello.

I added tests for this but they will only run in an "enable-zlib" build. This means they won't get run by travis, but any problems should be picked up by run-checker.

### CODE REVIEW

File: ssl/statem/statem_clnt.c [hunk 1]
```diff
@@ -1105,7 +1105,9 @@ int tls_construct_client_hello(SSL *s, WPACKET *pkt)
         return 0;
     }
 #ifndef OPENSSL_NO_COMP
-    if (ssl_allow_compression(s) && s->ctx->comp_methods) {
+    if (ssl_allow_compression(s)
+            && s->ctx->comp_methods
+            && (SSL_IS_DTLS(s) || s->s3->tmp.max_ver < TLS1_3_VERSION)) {
```
Comment (levitte) [inline ssl/statem/statem_clnt.c]: Would it make sense to add those new conditions in `ssl_allow_compression()`?

Comment (mattcaswell) [inline ssl/statem/statem_clnt.c]: Well, I did think about that but ssl_allow_compression() is called by both the client and the server and the conditions are slightly different on each side. In the client case this is called before TLSv1.3 is actually selected (we are constructing the ClientHello so version negotiation hasn't happened yet). These conditions check whether the ClientHello is compatible with TLSv1.3. On the server side we check whether TLSv1.3 has actually been selected.

File: test/ssl_test_ctx.c [hunk 2]
```diff
@@ -286,6 +286,29 @@ const char *ssl_session_ticket_name(ssl_session_ticket_t server)
                      server);
 }
 
+/* CompressionExpected */
+
+static const test_enum ssl_compression[] = {
```
Comment (richsalz) [inline test/ssl_test_ctx.c]: why do we have separate types for all these booleans?  why not just a test_on test_off ? 
okay, it's not "all" just one other one.  but it should be merged.

Comment (mattcaswell) [inline test/ssl_test_ctx.c]: Good question. Apparently there already is a built-in boolean type that I should have been using.

**Comment by mattcaswell** [comment]
Pushed new commit to address review comments.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**